server/src-lib/Hasura/GraphQL/Transport/WebSocket.hs

```diff
@@ -546,6 +549,8 @@ onClose
 onClose logger lqMap wsConn = do
   logWSEvent logger wsConn EClosed
   operations <- liftIO $ STM.atomically $ ListT.toList $ STMMap.listT opMap
+  -- TODO We race several large STM transactions, with the possibility of livelock.
```

Comment:
I agree, no point in running them concurrently but I don't see how a livelock can happen, given that each of these transactions only touch a single `TVar`.

Reply:
Hm not clear to me what you mean (I'm looking at body of `removeLiveQuery` in which we have racing `lookupfrommap ... stmstuff... deletefrommap`), would be interesting to look into it further, but just making it serial seems more pragmatic :+1:

Reply:
Ignore my previous comment! I completely misunderstood what `removeLiveQuery` is doing. So the reason why this is concurrent is because of the assumption that each websocket connection would most likely have subscriptions that map to different pollers than the same, in such case, executing `removeLiveQuery` concurrently is better because `STMMap` can handle such cases really well. 

>  A transaction operating on the map will only ever have to retry if another transaction is operating on the same key at the same time.

I'm not particular about this at all, we can execute it serially too. I think the best solution would be to group the queries that need to be deleted based on the `PollerKey`, each such group can be processed concurrently (in each group we delete the queries sequentially). Regardless, we'll still have contention from other closing connections.